convert/convert.go

```diff
@@ -318,321 +337,314 @@ func WithChunksCompression(compressionOpts ...schema.CompressionOpts) ConvertOpt
+	if len(allSeries) == 0 {
+		return 0, nil, nil
 	}
-	return err.ErrorOrNil()
-}
 
-func (rr *TsdbRowReader) Schema() *schema.TSDBSchema {
-	return rr.tsdbSchema
-}
-
-func sortedPostings(ctx context.Context, indexr tsdb.IndexReader, mint, maxt int64, sortedLabels ...string) index.Postings {
-	p := tsdb.AllSortedPostings(ctx, indexr)
-
-	if len(sortedLabels) == 0 {
-		return p
-	}
+	slices.SortFunc(allSeries, compareBlockSeriesBySortedLabelsFunc(opts.sortedLabels))
 
-	type s struct {
-		ref    storage.SeriesRef
-		idx    int
-		labels labels.Labels
+	// Count how many unique series will exist after merging across blocks.
+	uniqueSeriesCount := 1
+	for i := 1; i < len(allSeries); i++ {
+		if labels.Compare(allSeries[i].labels, allSeries[i-1].labels) != 0 {
+			uniqueSeriesCount++
+		}
 	}
-	series := make([]s, 0, 128)
-	chks := make([]chunks.Meta, 0, 128)
 
-	scratchBuilder := labels.NewScratchBuilder(10)
-	lb := labels.NewBuilder(labels.EmptyLabels())
-	i := 0
-P:
-	for p.Next() {
-		scratchBuilder.Reset()
-		chks = chks[:0]
-		if err := indexr.Series(p.At(), &scratchBuilder, &chks); err != nil {
-			return index.ErrPostings(fmt.Errorf("unable to expand series: %w", err))
-		}
-		hasChunks := slices.ContainsFunc(chks, func(chk chunks.Meta) bool {
-			return mint <= chk.MaxTime && chk.MinTime <= maxt
-		})
-		if !hasChunks {
-			continue P
+	// Divide rows evenly across shards to avoid one small shard at the end;
+	// use floats & round up so integer division does not cut off the remainder series.
+	totalShards := int(math.Ceil(float64(uniqueSeriesCount) / float64(opts.numRowGroups*opts.rowGroupSize)))
+	rowsPerShard := int(math.Ceil(float64(uniqueSeriesCount) / float64(totalShards)))
+
+	// For each shard index i, shardSeries[i] is a map of blockIdx -> []series.
+	shardSeries := make([]map[int][]blockSeries, totalShards)
+	for i := range shardSeries {
+		shardSeries[i] = make(map[int][]blockSeries)
+	}
+
+	shardIdx := 0
+	allSeriesIdx := 0
+	for shardIdx < totalShards {
+		seriesToShard := allSeries[allSeriesIdx:]
+
+		// First series in a shard will always be unique.
+		uniqueCount := 1
+		shardSeries[shardIdx][seriesToShard[0].blockIdx] = append(
+			shardSeries[shardIdx][seriesToShard[0].blockIdx], seriesToShard[0],
+		)
+		allSeriesIdx++
+
+		// Add series into shard, counting unique series until we reach rowsPerShard.
+		// If multiple blocks have series with the same labelset,
+		// all matching series are added to the shard but only counted once towards the row limit;
+		// they will be merged by the mergeChunkSeriesSet as they are iterated during the write.
+		for i := 1; i < len(seriesToShard); i++ {
+			prev := seriesToShard[i-1]
+			curr := seriesToShard[i]
+			shardSeries[shardIdx][curr.blockIdx] = append(shardSeries[shardIdx][curr.blockIdx], curr)
+
+			if labels.Compare(curr.labels, prev.labels) != 0 {
+				uniqueCount++
+			}
+			allSeriesIdx++
+			if uniqueCount >= rowsPerShard {
+				break
+			}
```

Comment:
I might be missing it cos the logic is not trivial, but: at this point when you've determined you have enough rows, couldn't it be that the next series has the same label set and hence we'd need to merge it with the current one (and would not count towards the unique series count)?

On the other hand, maybe we are OK with having a series split in two rows in different shards as long as we know it when querying, though I don't know if that was a deliberate decision. Bringing it to attention just in case.

Reply:
Fixed and made sure this was covered in the sorted labels test